StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1649FileNameMustMatchTypeName.cs

```diff
@@ -0,0 +1,123 @@
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using Microsoft.CodeAnalysis.Text;
+    using StyleCop.Analyzers.Settings.ObjectModel;
+
+    /// <summary>
+    /// The name of a C# code file does not match the first type declared in the file. For generics
+    /// that are defined as Class1&lt;T&gt; the name of the file needs to be Class1{T}.cs or Class1`1
+    /// depending on the fileNamingConvention setting.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class SA1649FileNameMustMatchTypeName : DiagnosticAnalyzer
+    {
+        /// <summary>
+        /// The ID for diagnostics produced by the <see cref="SA1649FileNameMustMatchTypeName"/>
+        /// analyzer.
+        /// </summary>
+        public const string DiagnosticId = "SA1649";
+
+        /// <summary>
+        /// The key used for passing the expected file name to the code fix provider.
+        /// </summary>
+        internal const string ExpectedFileNameKey = "ExpectedFileName";
+
+        private static readonly LocalizableString Title = new LocalizableResourceString(nameof(DocumentationResources.SA1649Title), DocumentationResources.ResourceManager, typeof(DocumentationResources));
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(DocumentationResources.SA1649MessageFormat), DocumentationResources.ResourceManager, typeof(DocumentationResources));
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(DocumentationResources.SA1649Description), DocumentationResources.ResourceManager, typeof(DocumentationResources));
+        private static readonly string HelpLink = "https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1649.md";
+
+        private static readonly DiagnosticDescriptor Descriptor =
+            new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, AnalyzerCategory.DocumentationRules, DiagnosticSeverity.Warning, AnalyzerConstants.EnabledByDefault, Description, HelpLink);
+
+        /// <inheritdoc/>
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+            ImmutableArray.Create(Descriptor);
+
+        /// <inheritdoc/>
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(HandleCompilationStart);
+        }
+
+        private static void HandleCompilationStart(CompilationStartAnalysisContext context)
+        {
+            context.RegisterSyntaxTreeActionHonorExclusions(HandleSyntaxTreeAction);
+        }
+
+        private static void HandleSyntaxTreeAction(SyntaxTreeAnalysisContext context)
+        {
+            var syntaxRoot = context.Tree.GetRoot(context.CancellationToken);
+            var settings = context.GetStyleCopSettings();
```

Comment:
:bulb: This can be moved down so it's only accessed if required. For maximum efficiency, consider an approach like SA1600 uses (a nested `Analyzer` class which holds analyzer state) to ensure the settings file is only requested once per compilation.